generators/generate.py

```diff
@@ -0,0 +1,207 @@
+#!/usr/bin/python
+"""
+Generates exercise_test.py files for tests with canonical data.
+"""
+import json
+import re
+import logging
+from argparse import ArgumentParser
+from pathlib import Path
+from itertools import repeat
+from string import punctuation, whitespace
+
+# 3rd party imports:
+from jinja2 import Environment, PackageLoader, select_autoescape, TemplateError
+from yapf.yapflib.yapf_api import FormatFile, style
+
+logging.basicConfig(format="%(levelname)s: %(message)s")
+LOGGER = logging.getLogger(__name__)
+
+
+def replace_all(string, chars, rep):
+    """
+    Replace any char in chars with rep, reduce runs and strip terminal ends.
+    """
+    trans = str.maketrans(dict(zip(chars, repeat(rep))))
+    return re.sub("{0}+".format(re.escape(rep)), rep,
+                  string.translate(trans)).strip(rep)
+
+
+def to_snake(string):
+    """
+    Convert pretty much anything to snake_case.
+    """
+    clean = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", string)
+    clean = re.sub("([a-z0-9])([A-Z])", r"\1_\2", clean).lower()
+    return replace_all(clean, whitespace + punctuation, "_")
+
+
+def to_camel(string):
+    """
+    Convert pretty much anything to CamelCase.
+    """
+    return "".join(w.title() for w in to_snake(string).split("_"))
+
+
+def is_error(case):
+    """
+    Determines if the given case is an error.
+    """
+    expect = case.get("expected")
+    return expect and isinstance(expect, dict) and "error" in expect
+
+
+def has_error(data):
+    """
+    Check if any case in the data is an error.
+    """
+    if "cases" in data:
+        for case in data["cases"]:
+            if has_error(case):
+                return True
+        return False
+    return is_error(data)
+
+
+def get_properties(data):
+    """
+    Get the list of the unique property values accessed in all tests.
+    """
+
+    def _get_properties(obj):
+        if "cases" in obj:
+            for case in obj["cases"]:
+                for prop in get_properties(case):
+                    yield prop
+            return
+        yield obj["property"]
+
+    return sorted(set(_get_properties(data)))
+
+
+def format_input(case):
+    """
+    Format the case's input values for use in a function call.
+    """
+    raw = case["input"]
+    if len(raw) == 1:
+        return repr(list(raw.values())[0])
+    if any(k.isdigit() for k in raw):
+        return repr(raw)
+    return ", ".join("{0}={1!r}".format(*i) for i in raw.items())
+
+
+def format_expect(case):
+    """
+    Format the case's expected values for use in a function call.
+    """
+    if is_error(case):
+        return repr(case["expected"]["error"])
+    return repr(case["expected"])
+
+
+def main():
```

Comment:
I highly recommend the following modification:

```Python
def main(args=None):  # where args may be a list of str
...
	parser.parse_args(args)  # uses sys.argv if args is None
```

This will make automating testing of generated test suites much easier.

Reply:
Interesting, hadn't thought of that use case. Fixed and pushed.